external-crates/move/crates/move-compiler/src/typing/deprecations.rs

```diff
@@ -0,0 +1,765 @@
+            | T::LValue_::BorrowUnpack(_, mident, dname, ty_params, fields)
+            | T::LValue_::Unpack(mident, dname, ty_params, fields) => {
+                if let Some(deprecation) = self.deprecated_type(mident, dname) {
+                    self.register_deprecation_warning(deprecation.clone(), dname.loc());
+                }
+
+                for (mident, name, loc) in
+                    ty_params.iter().flat_map(qualified_datatype_name_of_type)
+                {
+                    if let Some(deprecation) = self.deprecated_type(mident, name) {
+                        self.register_deprecation_warning(deprecation.clone(), loc);
+                    }
+                }
+
+                for (_, _, (_, (ty, lval))) in fields.iter() {
+                    self.handle_lvalue(lval);
+                    for (mident, name, loc) in qualified_datatype_name_of_type(ty) {
+                        if let Some(deprecation) = self.deprecated_type(mident, name) {
+                            self.register_deprecation_warning(deprecation.clone(), loc);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    // Process all the deprecation warnings and add them to the environment.
+    fn process_and_add_deprecation_warnings(self) {
+        let attr_deprecation = |location| match location {
+            AttributePosition::Module => TypeSafety::DeprecatedModule,
+            AttributePosition::Constant => TypeSafety::DeprecatedConstant,
+            AttributePosition::Struct => TypeSafety::DeprecatedStruct,
+            AttributePosition::Enum => TypeSafety::DeprecatedEnum,
+            AttributePosition::Function => TypeSafety::DeprecatedFunction,
+            AttributePosition::AddressBlock
+            | AttributePosition::Use
+            | AttributePosition::Friend
+            | AttributePosition::Spec => unreachable!(
+                "ICE: Deprecations are not allowed in '{}'s and should have already been checked",
+                location
+            ),
+        };
+
+        for (DeprecationId { deprecation, .. }, locs) in self.deprecation_warnings {
+            let mut locs: Vec<_> = locs
+                .into_iter()
+                .flat_map(|(position, locs)| {
+                    Self::merge_locations(locs.into_iter().collect())
+                        .into_iter()
+                        .map(move |loc| (position, loc))
+                })
+                .collect();
+            locs.sort_by_key(|(_, loc)| *loc);
+            locs.reverse();
+
+            let initial_diag = locs.pop().expect("ICE: locs should not be empty");
+
+            let attr_deprecation = attr_deprecation(deprecation.location);
+            let location_string = match deprecation.location {
+                AttributePosition::Module => format!(
+                        "This {} is deprecated since the whole module that it is declared in is marked deprecated",
```

Comment:
I found this hard to read in the case of aliases and methods. So I would suggest something like 
```suggestion
                        "The {position} {name} is deprecated{module_msg}"
```
Where 
```
let module_msg = match deprecation.location {
    AttributePosition::Module => ". It is deprecated since its whole module is marked deprecated",
    _ => "",
};
```